client-common/src/key/private_key.rs

```diff
@@ -1,15 +1,73 @@
+use crate::Transaction;
+use chain_core::tx::TransactionId;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 use rand::rngs::OsRng;
 use secp256k1::schnorrsig::{schnorr_sign, SchnorrSignature};
 use secp256k1::{recovery::RecoverableSignature, Message, PublicKey as SecpPublicKey, SecretKey};
 use zeroize::Zeroize;
 
+use crate::storage::HardwareWallet;
 use crate::{ErrorKind, PublicKey, Result, ResultExt, SECP};
 
+/// location for different types of private key
+#[derive(Clone)]
+pub enum PrivateKeyLocation {
+    /// normal private key
+    Local(PrivateKey),
+    /// external hardware wallet as a private key
+    External(HardwareWallet),
+}
+
+/// a object acts like a private key should impl the trait
+pub trait PrivateKeyAction: Sync + Send {
```

Comment:
there's not much of a point for trait if all methods take directly `PrivateKeyLocation` instead of generic `T: PrivateKeyAction`. Either add a test and make it generic -- e.g. something like
```
pub enum PrivateKeyLocation<T: PrivateKeyAction> {
     /// normal private key
     Local(PrivateKey),
     /// external hardware wallet as a private key
      External(T),
  }
```

or just add plain methods directly on `PrivateKeyLocation`, so that the client code that takes `PrivateKeyLocation` can just do `.sign(...)` instead of always pattern matching on its variants